File: 200. Number of Islands/200. Number of Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        struct Point {
            int h;
            int w;
            Point (int h, int w) : h(h), w(w) {}
            Point operator+ (Point& p) const {
                return Point(h + p.h, w + p.w);
            }
        };
        
        array<Point, 4> directions = {
            Point(1, 0),
            Point(0, 1),
            Point(-1, 0),
            Point(0, -1),
        };
        
        int height = grid.size();
        if (height == 0) {
            return 0;
        }
        int width = grid[0].size();
        
        vector<Point> lands;
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                if (grid[i][j] != '0') {
                    lands.push_back(Point(i, j));
                }
            }
        }
        
        int landCount = 0;
        queue<Point> q;
        for (Point& point : lands) {
            if (grid[point.h][point.w] == '1') {
                
                q.push_back(point);
                grid[point.h][point.w] = '0';
                landCount++;
                
                while (q.size() > 0) {
                    Point& p = q.front();
                    
                    for (Point direction : directions) {
                        Point newPoint = direction + p;
                        
                        int h = newPoint.h;
                        int w = newPoint.w;
                        if (0 <= h && h < height && 
                            0 <= w && w < width && 
                            grid[h][w] == '1') {
                            
                            grid[h][w] = '0';
                            q.push_back(Point(h, w));
                        }
                    }

                    q.pop_front();
                }
            }
        }
        return landCount;
    }
};

File: 200. Number of Islands/200. Number of Islands.rs
use std::collections::VecDeque;
impl Solution {
    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let mut grid = grid;
        let mut res = 0;
        for i in 0..m {
            for j in 0..n {
                if grid[i][j] == '1' {
                    res += 1;
                    Self::delete_island(&mut grid, i, j);
                }
            }
        }
        return res;
    }
    fn delete_island(mut grid: &mut Vec<Vec<char>>, i: usize, j: usize) {
        let m = grid.len();
        let n = grid[0].len();
        let mut deq: VecDeque<(usize, usize)> = VecDeque::new();
        deq.push_back((i, j));
        while let Some((i, j)) = deq.pop_front() {
            if grid[i][j] == '1' {
                grid[i][j] = '0';
                if 0 < i {
                    deq.push_back((i - 1, j));
                }
                if i + 1 < m {
                    deq.push_back((i + 1, j));
                }
                if 0 < j {
                    deq.push_back((i, j - 1));
                }
                if j + 1 < n {
                    deq.push_back((i, j + 1));
                }
            }
        }
    }
}
